### PR TITLE
Add `all-features` flag to add ability to enable all supported feature conformance tests.

### LINKED ISSUE #1638: Add ability to enable all Supported Feature Conformance tests
**What would you like to be added**:
A way to enable all core and extended conformance tests without explictly specifying [each supported feature suite name ](https://github.com/kubernetes-sigs/gateway-api/blob/d950009b34786a2fda3f5e8837c2de31b54b483d/conformance/utils/suite/suite.go#L34)

**Why this is needed**:
To be able to run the complete and latest set of conformance tests in the implementation repo (e.g. in [envoy-gateway](https://github.com/envoyproxy/gateway/blob/8dd60b1a3aaa8a93504aae0efd4d50315a43bd96/test/conformance/conformance_test.go#L56)) w/o having to manually track which new suite was added to the gateway-api repo

Relates to https://kubernetes.slack.com/archives/CR0H13KGA/p1668094311024819

### PR DESCRIPTION
**Pull request by gyohuangxin**
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

It adds `all-features` flag to add ability to enable all supported feature conformance tests. For example:

```
go test ./conformance --gateway-class istio --debug --all-features
```

**Which issue(s) this PR fixes**:

Fixes #1638 

**Does this PR introduce a user-facing change?**:

```release-note
1. Add `all-features` flag to enable all supported feature conformance tests.
```

### CODE REVIEW

**Comment by gyohuangxin** [comment]
@robscott @arkodg Could you review this PR? Thanks.

File: conformance/utils/flags/flags.go
Comment (arkodg) [inline conformance/utils/flags/flags.go:32]: flag name lgtm !

File: conformance/utils/suite/suite.go
Comment (arkodg) [inline conformance/utils/suite/suite.go:82]: is there any way a contributor defines the suite name once ? I fear the contributor will add it the constant w/o adding it to this map, and the burden to ensure its in sync will fall on the maintainer

Comment (gyohuangxin) [reply conformance/utils/suite/suite.go:82]: Yes, you are right, developers need to modify two places after the constant changes. But I can't find a simple way to define the suite name once in this codebase. Or I can add a comment to notice developer/code reviewer to ensure they are in sync?
@robscott Can you help with some advice?

**Comment by gyohuangxin** [comment]
@robscott Could you review it? Thanks

File: conformance/utils/suite/suite.go
Comment (arkodg) [reply conformance/utils/suite/suite.go:82]: I can't think of a better solution rn, can you add a comment to ensure they are in sync, TIA !

Comment (gyohuangxin) [reply conformance/utils/suite/suite.go:85]: IMO, there is no native set structure in golang, do you mean the `map[SupportedFeature]struct{}`? Alsowe should pass `AllFeatures` to suite's `SupportedFeatures`, so we should make them the same type.

Comment (shaneutt) [reply conformance/utils/suite/suite.go:85]: Sets are available in apimachinery:

https://github.com/kubernetes/apimachinery/tree/master/pkg/util/sets

e.g.:

```go
import "k8s.io/apimachinery/pkg/utils/sets"

var AllFeatures = sets.NewString(
	SupportReferenceGrant,
	SupportTLSRoute,
	SupportHTTPRouteQueryParamMatching,
	SupportHTTPRouteMethodMatching,
	SupportHTTPResponseHeaderModification,
	SupportRouteDestinationPortMatching,
	SupportGatewayClassObservedGenerationBump,
	SupportHTTPRoutePortRedirect,
	SupportHTTPRouteSchemeRedirect,
	SupportHTTPRoutePathRedirect,
	SupportHTTPRouteHostRewrite,
	SupportHTTPRoutePathRewrite,
)
```

```go
if AllFeatures.Has(SupportReferenceGrant) {
```

Comment (shaneutt) [reply conformance/utils/suite/suite.go:85]: Yes the implementation ultimately uses a map, so the benefit here is for the developers which I think is a fair thing to trade for, however I don't consider this suggestion a blocker I'll let others weigh in.

Comment (gyohuangxin) [reply conformance/utils/suite/suite.go:85]: Understood, if others also want the benefit, I can implement it.

Comment (robscott) [reply conformance/utils/suite/suite.go:85]: Now that the k8s sets util supports generics, I think it would be best to move both this and AllFeatures over to use that approach.

**Review by robscott — COMMENTED** [review]
Thanks @gyohuangxin! Agree with moving to k8s sets util, but otherwise this LGTM.

**Review by arkodg — APPROVED** [review]
lgtm, thanks for adding this !

File: conformance/utils/suite/suite.go
Comment (gyohuangxin) [reply conformance/utils/suite/suite.go]: > Now that the k8s sets util supports generics, I think it would be best to move both this and AllFeatures over to use that approach.

@robscott @shaneutt I tried to move it to K8s sets utils: https://github.com/kubernetes-sigs/gateway-api/pull/1642/commits/b85174a868892238ff589643da738f492fda4e35, can you review if it is correct?

File: conformance/utils/suite/suite.go
Comment (shaneutt) [inline conformance/utils/suite/suite.go:99]: It should be possible to define these this way, right?
```suggestion
var StandardCoreFeatures = sets.New(
	SupportReferenceGrant,
)

// AllFeatures contains all the supported features and can be used to run all
// conformance tests with `all-features` flag.
//
// Note that the AllFeatures must in sync with defined features when the
// feature constants change.
var AllFeatures = sets.New(
	SupportReferenceGrant,
	SupportTLSRoute,
	SupportHTTPRouteQueryParamMatching,
	SupportHTTPRouteMethodMatching,
	SupportHTTPResponseHeaderModification,
	SupportRouteDestinationPortMatching,
	SupportGatewayClassObservedGenerationBump,
	SupportHTTPRoutePortRedirect,
	SupportHTTPRouteSchemeRedirect,
	SupportHTTPRoutePathRedirect,
	SupportHTTPRouteHostRewrite,
	SupportHTTPRoutePathRewrite,
}
```

Comment (gyohuangxin) [reply conformance/utils/suite/suite.go]: @shaneutt Yes, I've changed like this. Thanks.

File: conformance/utils/suite/suite.go
Comment (shaneutt) [inline conformance/utils/suite/suite.go:160]: nit: now that we're using a set, you could just do an `Insert(feature)` here for every element in the `StandardCoreFeatures`, as duplicates wont matter. Not necessary though.

**Review by shaneutt — APPROVED** [review]
Would like @robscott to give the final LGTM since he's been reviewing this too, but does LGTM.

**Review by shaneutt — APPROVED** [review]

**Comment by gyohuangxin** [comment]
@robscott Can you review it?

File: conformance/utils/suite/suite.go
Comment (robscott) [inline conformance/utils/suite/suite.go:246]: I think this we actually want this?
```suggestion
		if !suite.SupportedFeatures.Has(feature) {
```

**Review by robscott — COMMENTED** [review]
Thanks @gyohuangxin!

File: conformance/utils/suite/suite.go
Comment (gyohuangxin) [reply conformance/utils/suite/suite.go]: @robscott It's my mistake, updated.

**Review by shaneutt — APPROVED** [review]
/lgtm

**Pull request merged**
